libs/blocks/quiz/utils.js

```diff
@@ -239,6 +239,22 @@ const getNestedFragments = (resultResources, productCodes, fragKey) => {
   return fragArray;
 };
 
+/**
+ * Normalizes the metadata keys in the metadata object,
+ * cleaning them up and removing all but alphanumeric characters
+ */
+const normalizeKeys = (data) => {
+  const keys = Object.keys(data);
+  for (const key of keys) {
+    const newKey = key.match(/[a-zA-Z0-9]/g).join('');
+    if (key !== newKey) {
```

Comment:
I think it would be safer to add them to a new empty object rather than edit the one we are iterating over.

Reply:
Is there a specific danger or issue that you had in mind here? I am leaving keys alone that don't have issues with the match and only adding a new key and removing the old if the key has unsafe characters and needs to change. 

If I make a new object, the result is the same, though I would have to add keys that don't have issues to the new object. The returned data will be the new object and the original will just be discarded.

Would you feel less worried if I refrain from deleting any keys?

Reply:
Here's what the machine gods of OpenAI say about it:

Modifying keys in an object while iterating through them can lead to unexpected behavior and errors in JavaScript. When you iterate over the keys of an object using a loop like for...in or Object.keys(), you are essentially creating a snapshot of the keys at the beginning of the iteration.

If you add or remove keys during the iteration, it can affect the loop's behavior. For example, adding a new key might cause the loop to skip some keys or repeat others. Removing a key may cause the loop to skip or miss keys as well.

Here's an example to illustrate the issue:

javascript
Copy code
let myObject = { key1: 'value1', key2: 'value2', key3: 'value3' };

for (let key in myObject) {
  console.log(key); // key1, key2, key3

  // Modifying the object during iteration
  if (key === 'key2') {
    delete myObject.key1; // Removing a key
    myObject.newKey = 'newValue'; // Adding a new key
  }
}

console.log(myObject); // { key2: 'value2', key3: 'value3', newKey: 'newValue' }
In this example, the loop does not iterate over the newly added key (newKey), and it skips the deleted key (key1). This behavior can lead to unintended consequences and bugs in your code.

To avoid such issues, it's recommended not to modify the structure of the object while iterating over its keys. If you need to make changes, consider creating a new object or using a different approach to achieve your desired result.

Reply:
yes, this is from ChatGPT:
To improve the normalizeKeys function in JavaScript for better performance and clarity, consider the following modifications:

Use Object.entries and reduce: Instead of iterating with a for loop and modifying the original object, use Object.entries to get key-value pairs and reduce to build a new object. This approach avoids modifying the original object and is generally considered a more functional programming style.

Simplify Regular Expression: If the goal is to keep only alphanumeric characters in the keys, the regular expression is fine. However, ensure it meets your specific needs.

Check for Unnecessary Changes: Avoid creating a new key if it's identical to the original, which can save processing time.

Here's how the improved function might look:

javascript
Copy code
const normalizeKeys = (data) => {
  return Object.entries(data).reduce((newObj, [key, value]) => {
    const newKey = key.match(/[a-zA-Z0-9]/g)?.join('') || key;
    newObj[newKey] = value;
    return newObj;
  }, {});
};

Reply:
@echampio-at-adobe that makes sense, but I'm not actually iterating through the object here: I've taken a snapshot of the keys only using Object.keys, which gives an array of the keys, which is what I'm actually iterating through. I run a newKey through the match and if it's not the same as the base key, then update the object. 

Still think this is an issue, given that I'm looping through an array of the keys, rather than the object itself?

Reply:
I've adjusted this to just use a new object. If there's a difference, use the matched key in the new object, if there wasn't anything to change, just add it to the new object as is.